master/internal/api_master.go

```diff
@@ -25,3 +26,41 @@ func (a *apiServer) GetMaster(
 		Config:    configStruct,
 	}, err
 }
+
+func (a *apiServer) MasterLogs(
+	req *apiv1.MasterLogsRequest, resp apiv1.Determined_MasterLogsServer) error {
+	if err := grpc.ValidateRequest(
+		func() (bool, string) { return req.Limit >= 0, "Limit must be >= 0" },
+	); err != nil {
+		return err
+	}
+	total := a.m.logs.Len()
+	offset := int(req.Offset)
+	if req.Offset < 0 {
+		offset = total + offset
+		if offset < 0 {
+			offset = 0
+		}
+	}
+
+	limit := -1
+	if req.Limit != 0 {
+		limit = int(req.Limit)
+		if limit > total-offset {
+			limit = total - offset
+		}
+	}
+
+	for {
+		for _, log := range a.m.logs.Entries(offset, -1, limit) {
+			offset += 1
+			limit -= 1
+			if err := resp.Send(log.Proto()); err != nil {
+				return err
+			}
+		}
+		if !req.Follow || limit == 0 {
+			return nil
+		}
```

Comment:
If we tail the logs, we'll overflow `limit` eventually and after ~ 2 * math.MaxInt64 exit because `limit == 0` on line 62.

Reply:
Might should fix this by adding an
```
else if req.Follow {
   limit = -1
}
```
somewhere?